File: agent/library/src/main/java/org/matilda/MatildaAgent.java
package org.matilda;

import org.matilda.di.DaggerMatildaComponent;
import org.matilda.di.LoggerModule;
import org.matilda.di.MatildaComponent;
import org.matilda.di.MatildaConnectionModule;
import org.matilda.logger.Logger;

import java.io.IOException;

public class MatildaAgent {
    private final MatildaComponent mMatildaComponent;

    public MatildaAgent(MatildaConnection matildaConnection, Logger... loggers) {
        mMatildaComponent = DaggerMatildaComponent.builder()
                .matildaConnectionModule(new MatildaConnectionModule(matildaConnection))
                .loggerModule(new LoggerModule(loggers))
                .build();
        mMatildaComponent.destructionManager().addDestructor(matildaConnection::close);
    }

    public void run() {
        try {
            mMatildaComponent.messageListener().start();
        } catch (IOException e) {
            throw new RuntimeException(e);
        } finally {
            mMatildaComponent.destructionManager().destruct();
        }
    }
}

File: agent/library/src/main/java/org/matilda/MatildaAgent.kt
package org.matilda

import kotlinx.coroutines.runBlocking
import org.matilda.di.LoggerComponent
import org.matilda.di.MatildaComponent
import org.matilda.di.MatildaConnectionComponent
import org.matilda.di.create
import org.matilda.logger.Logger

class MatildaAgent(matildaConnection: MatildaConnection, vararg loggers: Logger) {
    private val mMatildaComponent: MatildaComponent

    init {
        mMatildaComponent = MatildaComponent::class.create(
            MatildaConnectionComponent::class.create(matildaConnection),
            LoggerComponent::class.create(loggers)
        )
        mMatildaComponent.destructionManager.addDestructor { matildaConnection.close() }
    }

    fun run() {
        runBlocking {
            try {
                mMatildaComponent.messageListener.start()
            } finally {
                mMatildaComponent.destructionManager.destruct()
            }
        }
    }
}
